ext/drreg/drreg.c

```diff
@@ -65,17 +65,66 @@
 
 #define PRE instrlist_meta_preinsert
 
-/* This should be pretty hard to exceed as there aren't this many GPRs */
-#define MAX_SPILLS (SPILL_SLOT_MAX + 8)
+/* This is an arbitrary hard-coded upper limit of how many slots drreg is able to track.
+ * This should accommodate all use-cases. Note, the client is responsible for reserving
+ * enough slots for its use.
+ */
+#define ARBITRARY_UPPER_LIMIT (SPILL_SLOT_MAX + DR_NUM_GPR_REGS + 1)
+#define MAX_SPILLS ARBITRARY_UPPER_LIMIT
 
-#define AFLAGS_SLOT 0 /* always */
+/* We choose the number of available slots for spilling SIMDs to arbitrarily match
+ * double their theoretical max number for a given build.
+ *
+ * Indirect spill area for SIMD is always allocated in TLS and therefore suitable for
+ * cross-app. Note that this is in contrast to GPRs, which requires allocated raw thread
+ * storage for cross-app spilling as DR slots are not guaranteed to preserve stored data
+ * in such cases.
+ */
+#define MAX_SIMD_SPILLS (DR_NUM_SIMD_VECTOR_REGS * 2)
+
+#ifdef X86
+/* Defines whether SIMD and indirect spilling is supported by drreg. */
+#    define SIMD_SUPPORTED
+#endif
+
+#ifdef SIMD_SUPPORTED
+#    define XMM_REG_SIZE 16
+#    define YMM_REG_SIZE 32
+#    define ZMM_REG_SIZE 64
+#    define SIMD_REG_SIZE ZMM_REG_SIZE
+#    define _IF_SIMD_SUPPORTED(x) , x
+#else
+#    define _IF_SIMD_SUPPORTED(x)
+/* FIXME i#3844: NYI on ARM */
+#endif
 
-/* We support using GPR registers only: [DR_REG_START_GPR..DR_REG_STOP_GPR] */
+#define AFLAGS_SLOT 0 /* always */
 
+/* Liveness states for gprs */
 #define REG_DEAD ((void *)(ptr_uint_t)0)
 #define REG_LIVE ((void *)(ptr_uint_t)1)
 #define REG_UNKNOWN ((void *)(ptr_uint_t)2) /* only used outside drmgr insert phase */
 
+#ifdef SIMD_SUPPORTED
+/* Liveness states for SIMD vector registers (not for mmx).
+ * Note that value order, i.e., SIMD_ZMM_DEAD > SIMD_YMM_DEAD > SIMD_XMM_DEAD,
+ * is important as drreg relies on it to reason over states.
+ */
+#    define SIMD_XMM_DEAD \
+        ((void *)(ptr_uint_t)0) /* first 16 bytes are dead, rest are live */
+#    define SIMD_YMM_DEAD \
+        ((void *)(ptr_uint_t)1) /* first 32 bytes are dead, rest are live */
+#    define SIMD_ZMM_DEAD \
+        ((void *)(ptr_uint_t)2) /* first 64 bytes are dead, rest are live */
+#    define SIMD_XMM_LIVE \
+        ((void *)(ptr_uint_t)3) /* first 16 bytes are live, rest are dead */
+#    define SIMD_YMM_LIVE \
+        ((void *)(ptr_uint_t)4) /* first 32 bytes are live, rest are dead */
+#    define SIMD_ZMM_LIVE \
+        ((void *)(ptr_uint_t)5) /* first 64 bytes are live, rest are dead */
```

Comment:
Hmm, nicer to have a regular clean enum for these values, and then a macro on entry and exit of the drvector to do the void* cast?